### ISSUE TITLE
built apk(or aab) size problem

### ISSUE THREAD

**Issue by ciwhiz** [google-home/sample-app-for-matter-android#82]
hi,
I am experimenting to apply the matter library to my app.
It's the same when I built it in my app, but also when I saw it in this sample app.
I have a problem with apk size. 

After building, the apk' size is too large, because libCHIPController.so file size is too large. 

When I build on m1 Mac, the libCHIPController.so file is only 115/114 (Bytes) in arm64-v8a/armeabi-v7a in Analyze APK
However, when built on an intell cpu PC, it occupies 54.1 / 52.8 (MB). The gap is huge ~..;
 
Basically, Build options are the same. In app build.gradle, only arm64-v8a/armeabi-v7a was turned on for ndk.abiFilter, 
and the jniLibs settings were set as follows.
packagingOptions {
        jniLibs {
            useLegacyPackaging false
        }
    }

how can I apply this library into my app, in the smallest possible capacity.

**Comment by THETCR** [comment]
Looking at the Git history it seems between version `1.2.0` and `1.2.1` something went horribly wrong.
The APK size grew from `97.3 MB` to `1.23 GB`. 

The Matter libraries itself weren't updated between those versions.
https://github.com/google-home/sample-app-for-matter-android/compare/v1.2.0...v1.2.1

Is the `Gradle` version the cause of this?

**Comment by lmmfranco** [comment]
Yeah, having the same problem. Is libchipcontroller.so supposed to be this big? (>300MB)

**Comment by nicelyjust** [comment]
@lmmfranco @ciwhiz Please use the NDK trip command
![image](https://user-images.githubusercontent.com/20015907/220847794-b67454ac-461e-48d6-9a7c-4ce9f9ef39e1.png)

**Comment by THETCR** [comment]
@nicelyjust 
That would mean the libraries are bloated with debug symbols, which is understandable for the Matter libraries.
But the changes for this project should not result in the differences in size of the APKs that are released here, it seems live code inclusion (three shaking) is not properly applied anymore.

**Comment by pierredelisle** [comment]
Not an expert on this. If anyone has specific details on how to fix, please share with the community. Thx!
(no clue what the NDK trip command is about...)

**Comment by ciwhiz** [comment]
Thank you guys.
I finally found the reason.
In conclusion, yeah~ @nicelyjust was right.
When creating the '.so' file, I didn't "strip" it properly.
Proper strip processing resulted in a normal sized '.so' files.

It was not ndk but rather gradle that confused me in finding the cause of this problem.
AGP (android gradle plugin) has a basic ndk version from version 3.6 or higher,
At this time, if the corresponding ndk version is installed on the build machine (PC) and the path is set, it can be assumed that AGP strips unstriped so files when packaging apk.
I don't know the detailed AGP internal operation.

**Comment by saitomea** [comment]
@ciwhiz so, just having ndk on path with no addiotional config should work as intended?

**Comment by pierredelisle** [comment]
Fixing where NDK is located when building the APK fixes this issue. 
The next release will therefore have an APK with a "proper" size.

We're working on a fix for https://github.com/google-home/sample-app-for-matter-android/issues/91, and will then cut that release.

**Comment by pierredelisle** [comment]
Might be longer than anticipated to get #91 fixed, so went ahead and released [1.4.0](https://github.com/google-home/sample-app-for-matter-android/releases/tag/v1.4.0). 
APK size is still big (105Mb), and we should definitely try to get it down more, but at least it's not over 1Gb anymore.

**Comment by ciwhiz** [comment]
@saitomea , 
First of all, yes.
@pierredelisle answered above, but he says he's fixing it,
If you set the ndk path well, the ".so files" will be packed with normal size.
Check whether the ndk path is well set with the ./ndk-build -v command.

**Comment by pierredelisle** [comment]
Fixed in PR #98.

**Comment by tnguyen-alarm** [comment]
>APK size is still big (105Mb), and we should definitely try to get it down more, but at least it's not over 1Gb anymore.

@pierredelisle I'm curious if you've made any progress on reducing the effect on the APK size any further. Or do you have any ideas that are worth trying?

This thread was already immensely helpful in pointing out the `strip` command so thanks for that!

**Comment by tnguyen-alarm** [comment]
We were able to reduce the size further for our app by building the release versions of the .so files rather than using the ones in CHIPTool. Each .so file is now around ~7 MB down from ~25 MB. Example script:
```
cd ~/connectedhomeip
 
./third_party/android_deps/set_up_android_deps.py
./third_party/java_deps/set_up_java_deps.sh
 
# Build Android release on each architecture
gn gen out/release/android_arm64 --args='target_os="android" target_cpu="arm64" is_debug=false android_sdk_root="XXX" android_ndk_root="YYY" enable_android_builds=true'
ninja -C out/release/android_arm64
```

**Comment by OlivierGre** [comment]
I have used 
`llvm-strip out/android-arm64-chip-tool/lib/jni/arm64-v8a/libCHIPController.so `
Which reduce the size to 27MB.
